### PR TITLE
RPC Records list filtering

### PR DESCRIPTION
**Pull request by mathemancer**
Related to #3690 

Adds filtering to the `records.list` RPC function.

**Technical details**

The filters implemented are derived from the current Front End code.

The `Filter` spec is documented, but some changes warrant extra attention.

First, `and` and `or` are now just filters like the rest. Each takes two operands. So, to get the SQL

```sql
WHERE id < 3 AND col1 > 'potato' AND col2 = 'potahto'
```

you submit

```json
{
    "type": "and",
    "args": [
        {
            "type": "and",
            "args": [
                {
                    "type": "lesser",
                    "args": [
                        {"type": "attnum", "value": 1},
                        {"type": "literal", "value": 3}
                    ]
                },
                {
                    "type": "greater",
                    "args": [
                        {"type": "attnum", "value": 2},
                        {"type": "literal", "value": "potato"}
                    ]
                }
            ]
        },
        {
            "type": "equal",
            "args": [
                {"type": "attnum", "value": 3},
                {"type": "literal", "value": "potahto"}
            ]
        }
    ]
}
```
In reality, this produces the SQL:

```sql
WHERE (id < 3) AND ((col1 > 'potato') AND (col2 = 'potahto'))
```
The reason for this change is that it lets the recursive function that parses the filter spec be memoryless. It's also much more flexible moving forward.

Another change is that the filter type is no longer a key, but is instead a value corresponding to the key `"type"` in each filter. This is much more typical, and (again) allows the parser to be memoryless.

Finally, and less importantly, we've changed `column_id` to `attnum` to conform with the other RPC functions.

## Checklist
 
- [X] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [X] My pull request targets the `develop` branch of the repository
- [X] My commit messages follow [best practices][best_practices].
- [X] My code follows the established code style of the repository.
- [X] I added tests for the changes I made (if applicable).
- [X] I added or updated documentation (if applicable).
- [X] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by seancolsen** [comment]
I've only read the PR description, but I wanted to chime in with some thoughts...

This looks nice! The JSON is basically an AST for a limited set of SQL expressions. Very cool! The way you're using the `type` field works really well with TypeScript idioms.

I have one minor nitpick that very well may not be worth addressing. I'm just mentioning this as food for thought, on the off chance that it might tip your mental calculus into making an adjustment.

Instead of:

```json
{
  "type": "lesser",
  "args": [
    {
      "type": "attnum",
      "value": 1
    },
    {
      "type": "literal",
      "value": 3
    }
  ]
}
```

I'd rather see something like:

```json
{
  "type": "lesser",
  "lhs": {
    "type": "attnum",
    "value": 1
  },
  "rhs": {
    "type": "literal",
    "value": 3
  }
}
```

Why? Because a "less-than" expression always requires exactly two arguments. Structuring the JSON fields _statically_ as `lhs` and `rhs` (instead of _dynamically_ as `args`) allows us to encode that requirement into the type system.

To be clear: I don't mean to suggest that we'd have `lhs` and `rhs` values for every type of expression — only for certain types.

Here's an example of a comparison that we might want to add in the future which would be structured differently:

```json
{
  "type": "in",
  "value": {
    "type": "attnum",
    "value": 1
  },
  "values": [
    {
      "type": "literal",
      "value": 3
    },
    {
      "type": "literal",
      "value": 4
    }
  ]
}
```

Overall this is minor though. Not worth slowing things down unless you happen to see other good reasons for making such a change too.

**Review by pavish — COMMENTED** [review]
@mathemancer The syntax looks great to me! I do not see any concerns.

Adding some thoughts towards @seancolsen's comment:

I still prefer the syntax in the PR description because it follows a uniform structure (with each node containing a `type` and `args/value`) which is much easier to parse and work with.

In this example:
```
{
  "type": "lesser",
  "lhs": {
    "type": "attnum",
    "value": 1
  },
  "rhs": {
    "type": "literal",
    "value": 3
  }
}
```

I'd prefer finding the `lhs` and `rhs` using the `type` key instead of having them as additional keys.

In the scenario on the PR description, one is a column and the other is a literal, so we can deduct that easily using `attnum` and `literal`, and it works great.

However there are scenarios where it might be required, say both arguments are columns, I'd prefer something like this instead:
```
{
  "type": "lesser",
  "args": [
    {
      "type": "lhs",
      "value": {
        "operand_type": "attnum",
        "operand": 4
      }
    },
    {
      "type": "rhs",
      "value": {
        "operand_type": "attnum",
        "operand": 3
      }
    }
  ]
}
```

This allows maximum flexibility w.r.t structure and even for us to nest the expressions if we ever want to, eg., comparing the output of one filter with another.

**Comment by mathemancer** [comment]
@seancolsen @pavish A couple of quick notes regarding your well-received commentary.

- My first attempt was similar to what @seancolsen suggested, but the irregularity of the form made things (much) more complex in the implementation of the parser. Basically, every key ends up being another case in the end, and there's kind of a combinatorial explosion because the function is recursive.
- Please note that the `args` array is _ordered_. If you put the `literal` before the `attnum` in the `args` array for a `lesser` object, you'll end up with something like `42 < mycolumn`. It seems like you both may not have noticed this.
- As for comparing outputs of filters or functions via nesting (as suggested by @pavish), the current form allows for that, and so does the underlying implementation. It's just that nesting is currently not that useful since all currently implemented functions or operators output boolean values. But, if (for example) you want a logical `implies` operator, i.e., with the truth table

  a | b | r
  |-|-|-|
  t | t | t
  t | f | f
  f | t | t
  f | f | t
 
  you can simply nest your filters inside a filter object of type `lesser_or_equal`. Not sure how useful that would be for a user though.


Finally, I should point out that the form, and the underlying parser, is flexible enough to allow for writing a whole bunch of arbitrary SQL. It doesn't actually know it's writing a "filter". I intend to use it for more general function calls in the future.
  
As a quick example, the `json_array_length_lesser_or_equal` filter is a bit redundant. We could easily implement a `json_array_length` 'filter' (really just a preproc function call at that point), and then the filter
  ```json
  {
      "type": "or",
      "args": [
          {
              "type": "lesser",
              "args": [
                  {
                      "type": "jsonb_array_length",
                      "args": [
                          {"type": "attnum", "value": 2}
                      ]
                  },
                  {"type": "literal", "value": 3}
              ]
          },
          {
              "type": "equal",
              "args": [
                  {
                      "type": "jsonb_array_length",
                      "args": [
                          {"type": "attnum", "value": 2}
                      ]
                  },
                  {"type": "literal", "value": 3}
              ]
          }
      ]
  }
  ```
  would produce the SQL
  ```sql
  WHERE (jsonb_array_length(col1::jsonb) < 3) OR (jsonb_array_length(col1::jsonb) = 3)
  ```
This filter will give the exact same output as `json_array_length_lesser_equal`. I didn't end up going that far on this, because I wanted to minimize scope (on both ends) by keeping the list of functions the same as it was for the REST version. If it would be useful to have any of these 'preproc' functions (`json_array_length` would be a main candidate) available by beta, just let me know. It's 10 minutes to implement, and 10 minutes to write a quick test.

You may notice that _many_ of our current filter functions could actually just be compositions or argument-swaps of other functions. In the long run, I think we should lean on that feature.

**Comment by mathemancer** [comment]
@seancolsen Please take a look at the decomposed functions. I ended up keeping the `lesser_or_equals` functions as well as `greater_or_equals`, since I think it's more convenient that way and people are used to having those operators available.

I also went ahead and made separate versions of `contains` or `starts_with` based on case sensitivity or insensitivity. 

Finally, I realized once I got into the docs that there's a `starts_with` function that doesn't allow any wildcards at all, saving us potential mistakes. Also the `strpos` function returns a truthy value if the substring is in the value and a falsy one otherwise. Since it also doesn't accept wildcards, I went with that for our 'contains' logic.

**Review by seancolsen — APPROVED** [review]
Wonderful!

**Review by Anish9901 — CHANGES_REQUESTED** [review]

File: db/sql/00_msar.sql [hunk 1]
```diff
@@ -3303,6 +3303,52 @@ SELECT val;
...
+  ('null', '(%s) IS NULL'),
+  ('not_null', '(%s) IS NOT NULL'),
+  -- string specific filters
+  ('contains_case_insensitive', 'strpos(lower(%s), lower(%s))::boolean'),
+  ('starts_with_case_insensitive', 'starts_with(lower(%s), lower(%s))'),
+  ('contains', 'strpos((%s), (%s))::boolean'),
+  ('starts_with', 'starts_with((%s), (%s))'),
+  -- json(b) filters and expressions
+  ('json_array_length', 'jsonb_array_length((%s)::jsonb)'),
+  ('json_array_contains', '(%s) @> (%s)'),
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply db/sql/00_msar.sql:3327]: The main goal was to allow this to be used on both `json` and `jsonb`. But, your point stands. I'll add typecasting to `json_array_contains`.

[the remark this answers is not in the document]
Comment (Anish9901) [reply db/sql/00_msar.sql]: That makes sense.

**Review by Anish9901 — APPROVED** [review]
Looks good now @mathemancer, Thanks!

**Pull request merged**